Cross-compiling of Python modules with and without distutils

chewi opened the pull request:
This probably won't be merged all in one go like this but it's mainly to get some initial feedback. I'm pretty happy with it now but I'm open to suggestions.

chewi commented:
Here's the list of packages I've had success with. The Waf-based ones will fail for you as I have extra hacks to deal with that.

app-backup/rdiff-backup
dev-python/apsw
dev-python/bcrypt
dev-python/cffi
dev-python/coverage
dev-python/cryptography
dev-python/html5-parser
dev-python/lxml
dev-python/markupsafe
dev-python/msgpack
dev-python/netifaces
dev-python/numpy
dev-python/psutil
dev-python/pyalsa
dev-python/pyblake2
dev-python/pycryptodome
dev-python/pycurl
dev-python/pynacl
dev-python/pyrsistent
dev-python/pyxattr
dev-python/regex
dev-python/simplejson
dev-python/sip
dev-vcs/bzr
dev-vcs/mercurial
games-util/nml
sys-apps/portage
sys-devel/distcc
sys-libs/talloc
sys-libs/tdb
sys-libs/tevent

mgorny commented on eclass/python-utils-r1.eclass:
```diff
@@ -211,9 +211,12 @@ _python_impl_matches() {
 # Example value:
 # @CODE
-# /usr/bin/python2.7
+# /var/tmp/portage/dev-python/pyblake2-1.2.3/temp/python2.7/bin/python2.7
```
I think this will break stuff. One of the points of PYTHON being the real path is that people actually hardcode the path given. If you change it, then I think we'll end up with having scripts referencing `#!/var/tmp/portage/dev-python/pyblake2-1.2.3/temp/python2.7/bin/python2.7` installed.

chewi replied:
It's true that I did not consider this but I believe our python-exec stuff prevents this from happening with shebangs. I grepped my entire test cross system and did not find a single instance of a path like that in any non-binary file. I suppose it could hypothetically appear in other places but I'll build the whole of dev-python to find them if that's what it takes. Unless you have a better idea? Having PYTHON point to the wrapper is not necessary for most packages, generally just the ones that don't use distutils, so we could tackle those individually but it would still be quite tedious.

mgorny replied:
It's not `dev-python/` I'm worried about, it's everything outside `dev-python/`. Especially cheapo build systems.

mgorny replied:
python-exec doesn't affect the shebangs in scripts themselves. They're preserved as-is and respected, in `/usr/lib/python-exec/*`.

mgorny replied:
(this makes me think of adding a QA check to Portage to make sure thingies don't install stuff with generic shebangs there)

mgorny replied:
Oh, and please test it on regular system as well ;-). I'm not really opposed to those changes (though I haven't reviewed them completely) but I'm afraid this is going to cause silent retroactive breakage of stuff.

chewi commented:
Fixed dev-python/sip. Still looking into the other issue.

chewi commented:
After realising just how many Python packages there are, I tried to be smarter and grep'd for ebuilds where this might break. I did find some interesting cases, including several that are already broken because they use `python_newexe` on scripts with shebangs like `#!/usr/bin/env python`. Some used sed to replace `${PYTHON}` with something else and one replaced the existing shebang with `${PYTHON}`.

While the odd ebuild may need fixing with (or without!) this change, I tried to mitigate most cases by dealing with it in `python_fix_shebang`. This involved reworking the function to work in a slightly different way that I believe is an improvement. See for yourself. The interesting commits are d7170b039f56b9570793a822bae54335a535603d and c7d3a01a4284e86895f57ca0285a112a94dfc421.

stefson commented:
Would you mind adding cython on the list?

chewi commented:
cython isn't in the `_PYTHON_ALL_IMPLS` list so we don't support it at all right now?

stefson commented:
I meant to ask for `dev-python/cython` :-)

chewi commented:
That is very much outside the scope of this pull request.

stefson commented:
k, thanks for clarification.

mgorny commented on eclass/python-utils-r1.eclass:
```diff
@@ -1012,38 +1027,63 @@ python_wrapper_setup() {
+			ln -s "${sysrootlib}/${EPYTHON}"/_sysconfigdata*.py "${workdir}"/lib/_sysconfigdata.py || die
+
+			# Use distutils/sysconfig from SYSROOT as parts of it have
+			# GENTOO_LIBDIR baked in at Python build time.
```
is GENTOO_LIBDIR the only issue here?

chewi replied:
For those particular files, yes, I have just diff'd them to check. Does that mean we can avoid the `PYTHONPATH` issue? Not quite. You still have to override the file pointed at by `_PYTHON_SYSCONFIGDATA_NAME`. This variable cannot be a full path. It also needs to be set via a wrapper because as soon as you set it within the Portage environment, Portage itself keels over. Maybe we could patch in support for a full path but I'm not sure how we'd prevent it from screwing with Portage.

I need to rebase this because your patch was reverted in the end.

mgorny replied:
Ok, so if we got rid of GENTOO_LIBDIR altogether will it reduce the needed patch or not at all?

chewi replied:
Only the following call to ln.

chewi commented:
I think this is good to go now. It hasn't changed much since my last big push a couple of months ago but I've made `PYTHON` only point to the wrapper when `SYSROOT` is not / to avoid unnecessary complexity. I cross-built the above list of packages again and also built them natively on a regular system to check.

I did have a very long think about whether there was another way to deal with the `_sysconfigdata.py` and `PYTHONPATH` issue to avoid having to rely on the wrapper. There really isn't. Even if you patched Python or did something clever with bind mounts, you need these overrides to only apply when Python is invoked in the context of the code being built, as opposed to being invoked by some external dependency or some Portage helper. Calling the wrapper via `PYTHON` or the `PATH` is the only effective way to do that.

chewi commented:
It occurs to me that this is slightly broken for prefix users although the existing version is broken too. `PYTHON` is primarily used for building, which means it should be defined using `BROOT` rather than `EPREFIX` but this value will sometimes find its way into shebangs, where the opposite is true. You also have to consider prefix bootstrapping, where IIRC Python should be executed from the target prefix rather than the build prefix. I've forgotten exactly how this works but I think it relies on the PATH. More on this soon.

chewi commented:
Apologies @stefson, I only now realise that I was totally confused about cython, thinking it was some kind of alternative Python implementation. No idea why. I'm pleased to report that it already cross-compiles with these changes.

stefson commented:
I applied this pullrequest as a patch against the portage tree, which is at `b6f034485fba47fc5e626da9fce75e80f4d3e8d0` , and got a few checksum warnings from portage:  
```
Calculating dependencies / * Digest verification failed:
 * /usr/portage/dev-libs/boost/boost-1.65.0.ebuild
 * Reason: Filesize does not match recorded size
 * Got: 12239
 * Expected: 12440
* Digest verification failed:
 * /usr/portage/sys-libs/tdb/tdb-1.3.16.ebuild
 * Reason: Filesize does not match recorded size
 * Got: 1381
 * Expected: 1332
 - * Digest verification failed:
 * /usr/portage/sys-libs/tevent/tevent-0.9.37.ebuild
 * Reason: Filesize does not match recorded size
 * Got: 1625
 * Expected: 1576
 \ * Digest verification failed:
 * /usr/portage/sys-libs/talloc/talloc-2.1.14.ebuild
 * Reason: Filesize does not match recorded size
 * Got: 2422
 * Expected: 2373
```

chewi commented:
Ebuild manifests are not included in the git repository but get added to the rsync repository later so patching like that would lead to this problem. A quick workaround is to use `FEATURES="-strict"`.

stefson commented:
Ok, I added the digest locally and indeed `dev-python/cython` now cross compiles just fine. 

`=dev-python/pillow-5.2.0` still fails, which is recent stable on arm. Would you mind to extend the fixups from 5.3.0 to it?

chewi commented:
These are not my packages and changing stable ebuilds in this way is bad form regardless. Please be patient or make the changes yourself.

TheChymera commented:
I was told that this might solve bug https://bugs.gentoo.org/670234

Alas, it does not: [build.log](https://gist.githubusercontent.com/TheChymera/6308a5129576b240ea64e33174c9063c/raw/60e1b88d7c495ad6affe8bb48d217708e78182c5/gistfile1.txt)

chewi commented:
@mgorny I've now addressed the points you raised that did not require further discussion. I await your feedback on the remaining comments and patches.

TheChymera commented:
@chewi this solves bug https://bugs.gentoo.org/670234 now: [log](https://gist.github.com/TheChymera/d1c220c900bc55c3b81bd75b218d8024)

chewi commented:
I've reduced the number of commits here by merging some of the standalone fixes and dealing with the Waf issue in the waf-utils.eclass instead.

@mgorny, I'm not trying to rush you but please could you let me know if you at least intend to provide further feedback so I'm not waiting for nothing.

mgorny commented:
Yes, I do but it might get delayed until llvm 8 is branched (next week).

stefson commented:
I tested a cross compile of `dev-libs/boost-1.65` at runtime, and build some stuff against it. The cross compile itself needs to have `boost-build` installed on the host, and it must be in sync. Otherwise its just fine. Thank you :-)

chewi commented:
To be honest, cross-compiling boost worked before, I have only fixed the Python bindings here.

andrey-utkin commented:
Just a word of support and confirmation.

I have just completed cross-building `@system` set and then sone more random stuff on python3_6-only system of default/linux/arm/17.0/armv6j profile on an up to date Gentoo tree with @chewi's patchset rebased.

The patchset is really helpful. I haven't encountered any build failures related to Python cross-compiling issues. No packages have any files in /usr/lib64 installed.

Thanks a lot to @chewi for doing this important work, and for his kind support to me!

stefson commented:
I have to say the patchset was really helpfull for cross-compiling a lot of python related ebuilds, portage and gemato, and so on, you name it. 

But today I ran into a regression because of it: I set up a armv7a-unknown-linux-gnueabihf cross-compiler, and populated the prefix with all of the dependencies needed for a cross-compile of firefox by `emerge-armv7a-unknown-linux-gnueabi -av firefox -o`

Now, with the patchset applied to the portage tree, the configure function of the firefox ebuild can't find a module called `_io` so it seems, making the config fail. 

Logfile: [firefox-esr failed configure.log.gz](https://github.com/gentoo/gentoo/files/2966566/firefox-esr.failed.configure.log.gz)

this log is of a full emerge of the same ebuild, from portage tree without your patch: 
[firefox-esr-cross-build.log.gz](https://github.com/gentoo/gentoo/files/2966570/firefox-esr-cross-build.log.gz)

chewi commented:
I can't look in detail right now but I can see that it is creating a virtual Python environment within `${WORKDIR}`, which doesn't bode well. I didn't know that the Firefox build used Python but I'm not surprised that something as large and complex as this fails. I'll see what I can do.

chewi commented:
My test system was configured as headless so I've had a few deps to trawl through. I've run into a few small EAPI 7-related problems on the way. Nearly there now. More tomorrow.

chewi commented:
I think I've figured it out but I need to rebuild Rust with ARM support before I can try a full build. I was thinking it would be nice if packages with LLVM target flags could enforce these when cross-compiling but it's awkward with virtual/rust in the middle.

stefson commented:
I have a slightly modified ebuild available for rust-std-armv7: https://github.com/stefson/gentoo/pull/6
Its best to use system-llvm with llvm:7, for it to work you'll have to apply the same patches from the dev-lang/rust ebuild to your systems llvm:7

stefson commented:
unfortunatly, I found another sort of regression of this patchset: any package which relies on python-config to find its PYTHON_INCLUDES is notoriously difficult to handle when cross-compiling, because the hosts output for `python-config --includes` is `-I/usr/include/python3.6m -I/usr/include/python3.6m` and the target one is the same. This makes the configure flake as it can't find python-headers. A good example for this behaviour is `dev-libs/gobject-introspection` 

```
checking for python script directory... ${prefix}/lib64/python3.6/site-packages
checking for python extension module directory... ${exec_prefix}/lib64/python3.6/site-packages
checking for headers required to compile python extensions... not found
configure: error: Python headers not found
```

Now, with your patchset, this error changes:   
```/usr/armv7a-unknown-linux-gnueabihf/tmp/portage/dev-libs/gobject-introspection-1.56.1/temp/python-wrappers/python3.6-sysroot/bin/python3.6
checking whether /usr/armv7a-unknown-linux-gnueabihf/tmp/portage/dev-libs/gobject-introspection-1.56.1/temp/python-wrappers/python3.6-sysroot/bin/python3.6 version is >= 2.7... yes
checking for /usr/armv7a-unknown-linux-gnueabihf/tmp/portage/dev-libs/gobject-introspection-1.56.1/temp/python-wrappers/python3.6-sysroot/bin/python3.6 version... 3.6
checking for /usr/armv7a-unknown-linux-gnueabihf/tmp/portage/dev-libs/gobject-introspection-1.56.1/temp/python-wrappers/python3.6-sysroot/bin/python3.6 platform... linux
checking for /usr/armv7a-unknown-linux-gnueabihf/tmp/portage/dev-libs/gobject-introspection-1.56.1/temp/python-wrappers/python3.6-sysroot/bin/python3.6 script directory... ${prefix}/lib/python3.6/site-packages
checking for /usr/armv7a-unknown-linux-gnueabihf/tmp/portage/dev-libs/gobject-introspection-1.56.1/temp/python-wrappers/python3.6-sysroot/bin/python3.6 extension module directory... ${exec_prefix}/lib/python3.6/site-packages
checking for headers required to compile python extensions... ./configure: line 15236: /usr/armv7a-unknown-linux-gnueabihf/tmp/portage/dev-libs/gobject-introspection-1.56.1/temp/python-wrappers/python3.6-sysroot/bin/python3.6-config: No such file or directory
not found
configure: error: Python headers not found
```

just to mention this for feedback, as it changes wrong behaviour into something I don't fully understand.

chewi commented:
Please don't call these regressions, it's not like these packages cross-compiled before. I also didn't claim that this was a magic wand that would work for every single package. You've flagged one of the biggest and one of the most cross-unfriendly packages in the tree.

With a one line fix to mozcoreconf-v6.eclass, I was able to get 5 minutes into the Firefox build but it failed on something libevent-related. I will investigate further.

gobject-introspection uses python-config, which sucks in itself, but it also assumes it is located at ${PYTHON}-config. If I create a wrapper called python3.6-config rather than just python-config and python3-config then it works. It only gets a bit further though because, as I said, gobject-introspection is one of the most cross-unfriendly packages going. Even if you manage to build it, any other package that tries to cross-build with it will fail miserably. Your best bet is to disable USE=introspection. If you need that flag, you're screwed, sorry. I've tried to work around this in cross-boss but it's not pretty.

stefson commented:
Sure, I chose a bad example. It's just that there are propably more build scripts that use `python-config` to determine cflags in their configure scripts, it was meant as a warning to hopefully help you with feedback. 

Also please don't take my choice of vocabulary for describing things going on for too literally, I'm not a native speaker and certainly not on par with your verbal skills. 

In any case, feel free to share the attempted patch for the mozcoreconf eclass if you need more feedback or exchange of thoughts.

gaboroszkar commented:
@chewi Do these changes also fix wrong includes for python module builds?
I have had problems with some `dev-python/*` packages include the wrong version of `/usr/include/python<VERSION>/*` files when cross-compiling, and I had to make changes to these manually. (It tried to use the build system includes.)

I really appreciate, that this Pull Request would fix the problem with cross-compile installs some `site-packages` into `/usr/lib64/*` instead of `/usr/lib/*` when cross-compiling from a 64 bit system to a 32 bit one! I'm all for these changes, very useful!

chewi commented:
I don't recall seeing the version problem specifically but it's a good idea to keep Python versions, among other things, in sync. I imagine these changes would fix that problem though.

I'm still going to try and get these changes in but there's something I need to do first.

gaboroszkar commented:
> I don't recall seeing the version problem specifically but it's a good idea to keep Python versions, among other things, in sync.

I was talking about the problem that includes were used from the build system, instead of the built system. There were no Python version problems. Sorry if I was misleading.

deorder commented:
When I run `/usr/armv7a-rpi2hs-linux-musleabihf/usr/bin/python2.7-config --includes` it returns:
```-I/usr/include/python2.7 -I/usr/include/python2.7```
which is not correct

When I run `/usr/armv7a-rpi2hs-linux-musleabihf/usr/bin/python3.7-config  --includes` it returns:
```-I/usr/armv7a-rpi2hs-linux-musleabihf/usr/include/python3.7m -I/usr/armv7a-rpi2hs-linux-musleabihf/usr/include/python3.7m```
as expected, apparently resolved using a shell script and using PWD

`libxml2`, `vim` and more ebuilds that are using `python2.7` with `USE="python"` fail with:
```error: #error "LONG_BIT definition appears wrong for platform (bad gcc/glibc config?).```

An issue which is also mentioned here:
https://bugs.gentoo.org/582130

I tried reverting the `libxml2` fix so that is will use the wrapper one instead:
From `${ROOT%/}${PYTHON}` to `${PYTHON}`, but I still get the same issue

This happens when cross compiling a 32bit environment on a 64bit environment.
Also mentioned on here at: https://github.com/gentoo/gentoo/pull/9822#issuecomment-508573247
Any idea how to fix this?

chewi commented:
@deorder, I get the impression you haven't actually tried my work here. To be clear, despite the age of this pull request, it hasn't been merged yet. Changes have been requested and related parts of PMS needs to be clarified first.

Although this work tackles many of the common issues, the `python2.7-config --includes` point you've raised may still be a problem but I haven't seen it cause any issues in practise.

deorder commented:
> @deorder, I get the impression you haven't actually tried my work here. To be clear, despite the age of this pull request, it hasn't been merged yet. Changes have been requested and related parts of PMS need to be clarified first.
> 
> Although this work tackles many of the common issues, the `python2.7-config --includes` point you've raised may still be a problem but I haven't seen it cause any issues in practise.

I am using everything that is includes in this pull request. Using your modifications I was able to cross-compile a RPi3 64bit environment without issues while before it did not. When I build a RPi2 32bit environment however I get the `error: #error "LONG_BIT definition appears wrong for platform (bad gcc/glibc config?)` during the build of some packages that use Python or are Python modules. By adding the modifications to `sitecustomize.py` I was able to build half of the packages that first did not, but still not everything. I am not criticizing your work here as it actually helps, but there is still something missing that cause 32bit packages to sometimes fail when build on a 64bit machine.

deorder commented:
Okay I found another method that only requires one file and some environment variables that seems to work without modifying any eclass files. I am still testing. For now it only works when I disable the portage sandbox.

andrew-aladev commented:
If you have an AMD threadripper - you can just rebuild everything using qemu. I am working on these [images here](https://github.com/andrew-aladev/test-images). But I have to build `dev-python/pyblake2` and `sys-apps/portage` somehow before chrooting inside.

I am going just to ignore new python cross compile eclass feature (for now) and return back to monkey patching of python itself. I am building cross image inside "host" native image, so I can do anything with this "host" python.

Please do not forget about this pull request, people are watching. Thank you.

chewi commented:
Don't worry, I haven't. I was looking at it last week. @mgorny said it was too complicated and he was right because trying to fix the bits that have broken in the meantime is making my head spin. I have a rough idea for a simpler approach.

andrew-aladev commented:
I've just created more less reliable workaround for now.

Python 3.6 already includes [right workaround](https://github.com/python/cpython/blob/master/Misc/python-config.sh.in): `prefix_real=$(installed_prefix "$0")`. It provides like `/usr/arm-unknown-linux-gnueabi` for any directory used by python. Python 2.7 should be fixed by monkey patching in the same way.

```
+-------------------------+
|                         |
|           +-----------+ |              +-----------+
|           |           | |              |           |
| native    | cross     | |    export    | cross     |
| container | container | | +----------> | container |
|           |           | |              | (+ qemu)  |
|           +-----------+ |              |           |
|                         |              +-----------+
+-------------------------+
```

I am patching this native container, it will be removed.

```
arm-unknown-linux-gnueabi-emerge -v1 dev-lang/python

find /usr/arm-unknown-linux-gnueabi/lib -maxdepth 1 -name ld* \
  -exec cp "{}" /lib/ \;
sed -i "s/export PYTHON=\${EPREFIX}\/usr\/bin\/\${impl}/export PYTHON=arm-unknown-linux-gnueabi-\${impl}/g" \
  /usr/portage/eclass/python-utils-r1.eclass
sed -i "s/\${EPYTHON:-python}/arm-unknown-linux-gnueabi-\${EPYTHON:-python}/g" \
  /usr/portage/eclass/distutils-r1.eclass
sed -i "s/esetup.py install \(.*\)/esetup.py install \1 --prefix=\"\/usr\"/g" \
  /usr/portage/eclass/distutils-r1.eclass
find /usr/portage/sys-apps/portage -maxdepth 1 -name portage-*.ebuild \
  -exec sed -i "s/\${D%\/}\${PYTHON_SITEDIR}/\${D%\/}\${PYTHON_SITEDIR#\${ROOT%\/}}/g" "{}" \; \
  -exec ebuild "{}" manifest \;

cp arm-unknown-linux-gnueabi-python3.6 /usr/bin/
ln -s /usr/bin/arm-unknown-linux-gnueabi-python3.6 /usr/bin/arm-unknown-linux-gnueabi-python3
ln -s /usr/bin/arm-unknown-linux-gnueabi-python3.6 /usr/bin/arm-unknown-linux-gnueabi-python
```

`arm-unknown-linux-gnueabi-python3.6` just runs installed python (with all required config for target platform) using qemu inside native container:

```
#!/bin/bash

quoted_args=""

for arg in "$@"; do
  quoted_args+=$(printf " %q" "$arg")
done

LD_LIBRARY_PATH="/usr/arm-unknown-linux-gnueabi/lib:/usr/arm-unknown-linux-gnueabi/usr/lib:/usr/arm-unknown-linux-gnueabi/usr/lib/gcc/arm-unknown-linux-gnueabi/9.2.0/" \
  eval /usr/arm-unknown-linux-gnueabi/usr/bin/python3.6 "$quoted_args"
```

After that `arm-unknown-linux-gnueabi-emerge -v1 sys-apps/portage` works perfect. You can emerge many python extensions using this workaround. The only problem is the usage of `PYTHON_` variables from `python_export (eclass/python-utils-r1.eclass)`. For example `pkg_preinst` from `sys-apps/portage`:

```
[[ -d ${D%/}${PYTHON_SITEDIR} ]] || die "${D%/}${PYTHON_SITEDIR}: No such directory"
```

`$D` is a local image path, `$PYTHON_SITEDIR` is the global python path including `/usr/arm-unknown-linux-gnueabi` prefix. `${D%/}${PYTHON_SITEDIR}` provides the right path only when `ROOT` equals `/`. So any mix of `$D` and `$PYTHON_` need to be fixed for this workaround.

```
${D%/}${PYTHON_SITEDIR#${ROOT%/}}
```

uis246 commented:
Any news?

chewi commented:
Not forgotten but no news, sorry.

chewi commented:
This is obsolete now, mainly thanks to the wonderful gpep517. Meson currently still builds native extensions with the wrong filename, but I've almost fixed that.
